pkg/diff/diff.go

```diff
@@ -262,68 +337,128 @@ func (t *TableDiff) getSourceTableChecksum(ctx context.Context, job *CheckJob) (
-		if err != nil {
-			return false, errors.Trace(err)
+	if filterByRand {
+		rand.Seed(time.Now().UnixNano())
+		r := rand.Intn(100)
+		if r > t.Sample {
+			chunk.State = ignoreState
+			return true, nil
 		}
+	}
+
+	chunk.State = checkingState
+	update()
 
-		eq, err := t.compareRows(sourceRows, targetRows, orderKeyCols)
+	if t.UseChecksum {
+		// first check the checksum is equal or not
+		equal, err = t.compareChecksum(ctx, chunk)
 		if err != nil {
 			return false, errors.Trace(err)
 		}
-
-		// if equal is false, we continue check data, we should find all the different data just run once
-		if !eq {
-			equal = false
+		if equal {
+			return true, nil
 		}
 	}
 
+	if t.UseChecksum && t.OnlyUseChecksum {
+		return false, nil
+	}
+
+	// if checksum is not equal or don't need compare checksum, compare the data
+	log.Info("select data and then check data", zap.String("table", dbutil.TableName(t.TargetTable.Schema, t.TargetTable.Table)), zap.String("where", chunk.Where), zap.Reflect("args", chunk.Args))
+
+	equal, err = t.compareRows(ctx, chunk)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
 	return equal, nil
 }
 
-func (t *TableDiff) compareRows(sourceRows map[string][]map[string]*dbutil.ColumnData, targetRows []map[string]*dbutil.ColumnData, orderKeyCols []*model.ColumnInfo) (bool, error) {
+func (t *TableDiff) compareChecksum(ctx context.Context, chunk *ChunkRange) (bool, error) {
+	// first check the checksum is equal or not
+	sourceChecksum, err := t.getSourceTableChecksum(ctx, chunk)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	targetChecksum, err := dbutil.GetCRC32Checksum(ctx, t.TargetTable.Conn, t.TargetTable.Schema, t.TargetTable.Table, t.TargetTable.info, chunk.Where, utils.StringsToInterfaces(chunk.Args), utils.SliceToMap(t.IgnoreColumns))
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if sourceChecksum == targetChecksum {
+		log.Info("checksum is equal", zap.String("table", dbutil.TableName(t.TargetTable.Schema, t.TargetTable.Table)), zap.String("where", chunk.Where), zap.Reflect("args", chunk.Args), zap.Int64("checksum", sourceChecksum))
+		return true, nil
+	}
+
+	log.Warn("checksum is not equal", zap.String("table", dbutil.TableName(t.TargetTable.Schema, t.TargetTable.Table)), zap.String("where", chunk.Where), zap.Reflect("args", chunk.Args), zap.Int64("source checksum", sourceChecksum), zap.Int64("target checksum", targetChecksum))
+
+	return false, nil
+}
+
+func (t *TableDiff) compareRows(ctx context.Context, chunk *ChunkRange) (bool, error) {
+	sourceRows := make(map[string][]map[string]*dbutil.ColumnData)
+	for i, sourceTable := range t.SourceTables {
+		rows, _, err := getChunkRows(ctx, sourceTable.Conn, sourceTable.Schema, sourceTable.Table, sourceTable.info, chunk.Where, utils.StringsToInterfaces(chunk.Args), utils.SliceToMap(t.IgnoreColumns), t.Collation)
```

Comment:
pre-calculate `utils.StringsToInterfaces(chunk.Args), utils.SliceToMap(t.IgnoreColumns)` before the `for` loop, reuse them both for source tables and target table?

Reply:
updated in https://github.com/pingcap/tidb-tools/pull/224/commits/5f204554c963e6817d528bc73f2d9a2fb7f105d6